wdl-format/src/format/task.rs

```diff
@@ -0,0 +1,421 @@
+//! A module for formatting elements in tasks.
+
+use anyhow::Result;
+use wdl_ast::v1::CommandPart;
+use wdl_ast::v1::CommandSection;
+use wdl_ast::v1::Decl;
+use wdl_ast::v1::RequirementsItem;
+use wdl_ast::v1::RequirementsSection;
+use wdl_ast::v1::RuntimeItem;
+use wdl_ast::v1::RuntimeSection;
+use wdl_ast::v1::TaskDefinition;
+use wdl_ast::v1::TaskItem;
+use wdl_ast::AstNode;
+use wdl_ast::AstToken;
+use wdl_ast::SyntaxElement;
+use wdl_ast::SyntaxKind;
+
+use super::comments::format_inline_comment;
+use super::comments::format_preceding_comments;
+use super::format_state::SPACE;
+use super::FormatState;
+use super::Formattable;
+use super::NEWLINE;
+
+impl Formattable for CommandSection {
+    fn format(&self, buffer: &mut String, state: &mut FormatState) -> Result<()> {
+        format_preceding_comments(&self.syntax_element(), buffer, state, false)?;
+
+        let command_keyword = self
+            .syntax()
+            .children_with_tokens()
+            .find(|c| c.kind() == SyntaxKind::CommandKeyword)
+            .expect("Command section should have a command keyword");
+        state.indent(buffer)?;
+        buffer.push_str("command");
+        format_inline_comment(&command_keyword, buffer, state, true)?;
+
+        if self.is_heredoc() {
+            let open_heredoc = self
+                .syntax()
+                .children_with_tokens()
+                .find(|c| c.kind() == SyntaxKind::OpenHeredoc)
+                .expect("Command section should have an open heredoc");
+            format_preceding_comments(&open_heredoc, buffer, state, true)?;
+            // Open braces should ignore the "+1 rule" followed by other interrupted
+            // elements.
+            if state.interrupted() {
+                state.reset_interrupted();
+                state.indent(buffer)?;
+            } else {
+                buffer.push_str(SPACE);
+            }
+            buffer.push_str("<<<");
+        } else {
```

Comment:
Long-term, will this be configurable? My inclination is rewrite `command` sections as `heredoc` style, unless the user specifically requests not to.

Reply:
It's not safe to coerce into heredoc style _until_ placeholder coercion is implemented (which I'm working on in another branch). If we switched braces to heredocs without swapping `${}` to `~{}` we could cause bugs. But we agreed earlier that placeholder coercion was a good idea, so heredoc coercion can follow that.